File: Vent/Boussole/Boussole/ViewController.m
//
//  ViewController.m
//  Boussole
//
//  Created by dodo-mac on 22/05/2015.
//  Copyright (c) 2015 dodo-mac. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()

@end


@implementation ViewController
@synthesize locationManager;



- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    locationManager=[[CLLocationManager alloc] init];
    locationManager.desiredAccuracy = kCLLocationAccuracyBest;
    locationManager.headingFilter = 1;
    locationManager.delegate = self;
    [locationManager startUpdatingHeading];
    
    if ([locationManager respondsToSelector:@selector(requestAlwaysAuthorization)])
    {
        [locationManager requestWhenInUseAuthorization];
        [locationManager requestAlwaysAuthorization];
    }
    [locationManager startUpdatingLocation];
    
    

    
    boussole = [[UIImageView alloc]initWithImage:[UIImage imageNamed:@"boussole"]];
    boussole.center = CGPointMake(self.view.frame.size.width*0.5, self.view.frame.size.height*0.5);
    [self.view addSubview:boussole];
    
    flecheWind = [[UIImageView alloc]initWithImage:[UIImage imageNamed:@"fleche"]];
    flecheWind.center = CGPointMake(boussole.frame.size.width*0.5, boussole.frame.size.height*0.5);
    flecheWind.transform = CGAffineTransformRotate(CGAffineTransformMakeScale(0.5, 0.5), M_PI/3);
    [boussole addSubview:flecheWind];
    
    labelSpeed = [[UILabel alloc]initWithFrame:CGRectMake(self.view.frame.size.width*0.25, self.view.frame.size.height-200, self.view.frame.size.width*0.75, 200)];
    labelSpeed.numberOfLines = 4;
    [self.view addSubview:labelSpeed];
    
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)locationManager:(CLLocationManager *)manager didUpdateHeading:(CLHeading *)newHeading{
    // Convert Degree to Radian and move the needle

    float newRad =  -newHeading.trueHeading * M_PI / 180.0f;
    [UIView animateWithDuration:0.5 animations:^{
        boussole.transform = CGAffineTransformMakeRotation(newRad);
    }];
    
}

- (void)locationManager:(CLLocationManager *)manager didFailWithError:(NSError *)error
{
    NSLog(@"didFailWithError: %@", error);
    UIAlertView *errorAlert = [[UIAlertView alloc]
                               initWithTitle:@"Error" message:@"Failed to Get Your Location" delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil];
    [errorAlert show];
}

- (void)locationManager:(CLLocationManager *)manager didUpdateToLocation:(CLLocation *)newLocation fromLocation:(CLLocation *)oldLocation
{
    coordinateLongitude = newLocation.coordinate.longitude;
    coordinateLatitude = newLocation.coordinate.latitude;
    NSLog(@"Location: Longitude %.8f Latitude %.8f", coordinateLongitude,coordinateLatitude);

    
    // Construction de l'url à récupérer
    NSURLRequest *request = [NSURLRequest requestWithURL:[NSURL URLWithString:[NSString stringWithFormat:@"http://api.openweathermap.org/data/2.5/weather?lat=%f&lon=%f",coordinateLatitude,coordinateLongitude]]];
   
    // execution de la requête et récupération du JSON via un objet NSData
    NSData *response = [NSURLConnection sendSynchronousRequest:request returningResponse:nil error:nil];
    
    // On récupère le JSON en NSString depuis la réponse
    NSString *json_string = [[NSString alloc] initWithData:response encoding:NSUTF8StringEncoding];
    NSError *jsonError;
    NSDictionary* dico = [NSJSONSerialization JSONObjectWithData:[json_string dataUsingEncoding:NSUTF8StringEncoding] options:NSJSONReadingMutableContainers error:&jsonError];
    
    if (dico)
    {
        NSLog(@"%@",[dico objectForKey:@"wind"]);
        
        float deg = [[[dico objectForKey:@"wind"]objectForKey:@"deg"]floatValue];
        [UIView animateWithDuration:0.5 animations:^{
            flecheWind.transform = CGAffineTransformRotate(CGAffineTransformMakeScale(0.4, 0.4), deg*M_PI/180.0f);
        }];
        labelSpeed.text = [NSString stringWithFormat:@"Speed: %.2f Km/h\nDegres: %.2f°\nTemperature: %.1f\nhumidité: %d%%",[[[dico objectForKey:@"wind"]objectForKey:@"speed"]floatValue],deg,[[[dico objectForKey:@"main"]objectForKey:@"temp"]floatValue]-275.16,[[[dico objectForKey:@"main"]objectForKey:@"humidity"]intValue]];
    }
    

}

@end

File: Vent/ViewController.swift
//
//  ViewController.swift
//  Vent
//
//  Created by dodo-mac on 22/05/2015.
//  Copyright (c) 2015 Dorian Rouby. All rights reserved.
//

import UIKit
import QuartzCore
import CoreLocation

class ViewController: UIViewController, CLLocationManagerDelegate {
    
    @IBOutlet weak var drapeauUIView: DrapeauView!
    @IBOutlet weak var textField: UITextField!
    
    var locationManager: CLLocationManager!
    var données: Weather!

    
    override func viewDidLoad() {
        super.viewDidLoad()
    }
    
    override func viewDidAppear(animated: Bool) {
        super.viewDidAppear(animated)
        locationManager = CLLocationManager()
        locationManager.delegate = self
        
        // Start compass
        locationManager.desiredAccuracy = kCLLocationAccuracyHundredMeters
        locationManager.headingFilter = 1
        locationManager.distanceFilter = 100 //mètres
        locationManager.requestWhenInUseAuthorization()
        locationManager.requestAlwaysAuthorization()
        locationManager.startUpdatingLocation()
        locationManager.startUpdatingHeading()
    }
    
    override func viewDidDisappear(animated: Bool) {
        locationManager.stopUpdatingLocation()
        locationManager.stopUpdatingHeading()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    // MARK: Location delegate
    func locationManager(manager: CLLocationManager, didUpdateHeading newHeading:CLHeading) -> () {
        // Convert Degree to Radian and move the needle
        let newRad: CGFloat =  CGFloat(-newHeading.trueHeading * M_PI / 180.0)
        UIView.animateWithDuration(0.5, animations: { () -> Void in
            if let fonctionne = self.données {
                self.drapeauUIView.transform = CGAffineTransformMakeRotation(newRad + CGFloat(self.données.windItem.directionRadians + M_PI))
            } else {
                println("Attention ! Les données ne sont pas encore initialisées")
                self.drapeauUIView.transform = CGAffineTransformMakeRotation(newRad)
            }
    })
    }
    
    func locationManager(manager: CLLocationManager!, didFailWithError error: NSError!) {
        NSLog("There was an error retrieving location:\n\(error)")
        if !CLLocationManager.headingAvailable() {
            println("\nTu utilises le simulateur, et tu as surement oublié d'activer la localisation...\n(Bolosse)\n\n")
        }
    }
    
    func locationManagerShouldDisplayHeadingCalibration(manager: CLLocationManager!) -> Bool {
        return true
    }
    
    func locationManager(manager: CLLocationManager!, didUpdateLocations locations: [AnyObject]!) {
        // on prend la dernière localisation
        let newLocation = locations.last as! CLLocation
        println("Latitude : \(newLocation.coordinate.latitude), longitude : \(newLocation.coordinate.longitude)")
        // Start weather fetch
        Weather.retrieveWeather(latitude: Float(newLocation.coordinate.latitude), longitude: Float(newLocation.coordinate.longitude)) { (resultat) -> Void in
            self.takeWeather(resultat)
        }
    }
    
    func takeWeather(météo: Weather) -> Void {
        données = Weather(latitude: météo.latitude, longitude: météo.longitude, windObject: météo.windItem)
        if !CLLocationManager.headingAvailable() {
            println("Attention ! Cet iPhone (?) n'a pas les informations sur l'orientation.\nEst-ce qu'il a une boussole au moins ?")
            UIView.animateWithDuration(0.3, animations: { () -> Void in
                self.drapeauUIView.transform = CGAffineTransformMakeRotation(CGFloat(self.données.windItem.directionRadians + M_PI))
            })
        }
        
        // Afficher le texte
        let fmt = NSNumberFormatter()
        fmt.maximumFractionDigits = 4
        fmt.minimumFractionDigits = 2
        let textFormatté = fmt.stringFromNumber(données!.windItem.speed)!
        textField.text = "\(textFormatté) m/s"
        textField.sizeToFit()
        
    }



}

